Meter.c

```diff
@@ -266,89 +315,571 @@ static void BarMeterMode_draw(Meter* this, int x, int y, int w) {
+      offset = offset * 2 + 1;
+   }
+   for (; offset < recordNumCells; offset++) {
+      record[offset].cell.itemIndex = UINT8_MAX;
+      record[offset].cell.details = 0x00;
    }
+   if (sum <= 0.0)
+      return;
+
+   int scaleExp = 0;
+   do {
+      double scaledTotal = ldexp(total, scaleExp);
+      if (!(scaledTotal <= DBL_MAX))
+         break;
+
+      // If there is a "top cell" which won't be completely filled, determine
+      // its color first.
+      double area = (sum / scaledTotal) * GRAPH_HEIGHT;
+      assert(area > 0.0 && area <= GRAPH_HEIGHT);
+
+      int topCell = (int)ceil(area) - 1;
+      uint8_t blanksAtTopCell = ((uint8_t)(((double)(topCell + 1) - area) * 4.0) * 2);
+      uint8_t topCellItem = (this->curItems - 1);
+      double topCellArea = 0.0;
+      if (blanksAtTopCell > 0 ||
+         ((!(this->total > 0.0)) &&
+         topCell == ((GRAPH_HEIGHT - 1) >> scaleExp) &&
+         topCell % 2 == 0)) {
+         topCellArea = (double)(8 - blanksAtTopCell) / 8.0;
+         sum = 0.0;
+         double bottom = 0.0;
+         double maxArea = 0.0;
+         for (uint8_t i = 0; i < this->curItems && sum < DBL_MAX; i++) {
+            if (!(this->values[i] > 0.0)) {
+               continue;
+            }
+            sum += this->values[i];
+            if (!(sum <= DBL_MAX))
+               sum = DBL_MAX;
+
+            double top = (sum / scaledTotal) * GRAPH_HEIGHT;
+            if (top > (double)topCell) {
+               // Find the item that occupies the largest area of the top cell
+               if ((top - bottom) > maxArea) {
+                  topCellItem = i;
+                  maxArea = top - ((maxArea > 0.0) ? bottom : (double)topCell);
+               }
+            }
+            bottom = top;
+         }
+      }
+      topCell += 1;
+
+      // Compute colors of the rest of the cells, using the largest remainder
+      // method (a.k.a. Hamilton's method).
+      // The Hare quota is (scaledTotal / GRAPH_HEIGHT).
+      int paintedHigh = topCell + (int)topCellItem + 1;
+      int paintedLow = 0;
+      double threshold = 0.5;
+      double thresholdHigh = 1.0;
+      double thresholdLow = 0.0;
+      // Tiebreak 1: Favor items with less number of cells. (Top cell is not
+      // included in the count.)
+      int cellLimit = topCell;
+      int cellLimitHigh = topCell;
+      int cellLimitLow = 0.0;
+      // Tiebreak 2: Favor items whose indices are lower.
+      uint8_t tiedItemLimit = topCellItem;
+      while (true) {
+         sum = 0.0;
+         double bottom = 0.0;
+         int cellsPainted = 0;
+         double nextThresholdHigh = 0.0;
+         double nextThresholdLow = 1.0;
+         int nextCellLimitHigh = 0;
+         int nextCellLimitLow = topCell;
+         uint8_t numTiedItems = 0;
+         for (uint8_t i = 0; i <= topCellItem && sum < DBL_MAX; i++) {
+            if (!(this->values[i] > 0.0)) {
+               continue;
+            }
```

Comment:
Yes. Because NaN values should not be drawn in the graph. I'm avoiding explicit `isnan` checks when the comparison operators can address NaNs for us (if IEEE semantics are followed).